### DISCUSSION TITLE
Animator Rollback

### DISCUSSION THREAD

**Discussion by RyanGarber** [FirstGearGames/FishNet#462] [Feature Request]
A seldom seen feature that would be killer is the ability to rollback an animator. All of the data is already stored in the NetworkAnimator when `forceAll` is true; perhaps you could apply it, (wait one frame?), and use it for lag-compensation. It seems, however, for truly accurate rollback, each state's applied `normalizedTime` would have to be offset by latency. Perhaps this could be an option that is exposed?

**Comment by duyca** [comment]
i got fully animancer support CPS. i highly recommend you should switch animator to animancer. it easy to you and no more graph nightmare.

**Reply by shadesbelow** [reply]
Hi duyca, are you able to share your solution for using Animancer with FishNet?

**Reply by duyca** [reply]
Hi, it been awhile when i switch to other networking solution. Here is all files for animancer and cps support i use for old project. Not sure it will work for present Fishnet version.
[CustomKeyAnimancerComponent.txt](https://github.com/user-attachments/files/17159634/CustomKeyAnimancerComponent.txt)
[LobbyCharacterPrediction.txt](https://github.com/user-attachments/files/17159636/LobbyCharacterPrediction.txt)
[NetworkAnimancer.cs.txt](https://github.com/user-attachments/files/17159637/NetworkAnimancer.cs.txt)
Hope it will help.

**Reply by shadesbelow** [reply]
Wow, thanks so much for sharing this! I was afraid I wasn't going to be able to use Animancer with Fish Net

**Comment by FirstGearGames** [comment]
This has been considered for awhile but rolling back an animator would be **EXTREMELY** expensive and probably not even work 100% just because of how the animator behaves internally. There's no plans to add this in, and with the new prediction system you can just update animators locally on clients using state forwarding in the CSP script.
